src/bot/music.rs

```diff
@@ -171,40 +184,69 @@ impl MusicBot {
         Ok(())
     }
 
-    pub async fn add_audio(&mut self, url: String, user: String) -> anyhow::Result<()> {
-        match youtube_dl::get_audio_download_from_url(url, &self.logger).await {
-            Ok(mut metadata) => {
-                metadata.added_by = user;
-                info!(self.logger, "Found source"; "url" => &metadata.url);
+    pub async fn add_audio(&mut self, uri: String, user: String) -> anyhow::Result<()> {
+        let metadata = if uri.starts_with("file://") && self.music_root.is_some() {
+            let path = self.music_root.as_ref().unwrap().join(&uri[7..]);
+            let path = match path.canonicalize() {
+                Ok(p) => p,
+                Err(e) => {
+                    self.send_message(format!("Invalid path: {}", e)).await?;
+                    return Err(e)?;
+                }
+            };
 
-                self.playlist.push(metadata.clone());
+            // Make sure files outside of the root path can't be accessed
+            if !path.starts_with(self.music_root.as_ref().unwrap()) {
+                self.send_message(String::from("Invalid path")).await?;
+                return Err(anyhow!("Invalid path"));
+            }
 
-                if !self.player.is_started() {
-                    let entry = self.playlist.pop();
-                    if let Some(request) = entry {
-                        self.start_playing_audio(request).await?;
-                    }
-                } else {
-                    let duration = if let Some(duration) = metadata.duration {
-                        format!(" ({})", ts::bold(&humantime::format_duration(duration)))
-                    } else {
-                        format!("")
-                    };
-
-                    self.send_message(format!(
-                        "Added {}{} to playlist",
-                        ts::underline(&metadata.title),
-                        duration
-                    ))
-                    .await?;
+            let metadata = metadata_from_file(&path, &user);
+            metadata.unwrap_or_else(|| AudioMetadata {
+                uri: format!("file://{}", path.to_string_lossy()),
+                webpage_url: None,
+                title: path.file_name().unwrap().to_string_lossy().to_string(),
```

Comment:
file_name() returns option? what happens if i `!add file://`

Reply:
I will check if the path is a file before this.